adds some inline docs to hash constructors

KCErb opened the pull request:
As described in #6918, here's a little documentation for hash construction. 

I decided to nix the `initialize` docs since the other two `self.new` methods are wrappers to it. Let me know if that's misguided.

I'm open to adding in docs for other initialization methods (JSON, YAML), I would just need to actually learn about them (haven't used them myself) so it would slow this down a little.

r00ster91 commented:
Run `crystal tool format` to format your code and fix CircleCI.

KCErb commented:
@r00ster91 Thanks for the pointer, forgot that that runs on doc comments too!
@wontruefree Good point, I changed the examples a bit to address it.

straight-shoota commented on src/hash.cr:
```diff
@@ -23,10 +24,36 @@ class Hash(K, V)
+  # end
+  # new_hash.size # => 0
+  # new_hash["zero"] = 0
+  # new_hash["two"] = 2
```
These two lines don't really help illustrate the purpose. I think this would be a better example:
```cr
hash = Hash(String, Int32).new do |hash, key|
  hash[key] = key.size
end

hash.size        # => 0
hash["foo"]      # => 3
hash.size        # => 1
hash["bar"] = 10
hash["bar"]      # => 10
```

KCErb replied:
Sounds good to me. I was also thinking of borrowing the "go fish" example used in the Ruby docs. Does anyone want to comment on the value of *that* example as opposed to this one?

They also had one where you used the block feature to essentially use the hash as a cache for some kind of calculation that can be done using the key alone. But I wondered if that was too abstract.

straight-shoota commented on src/hash.cr:
````diff
@@ -23,10 +24,36 @@ class Hash(K, V)
 
+  # Creates a new empty `Hash` where the *default_value* is returned if key is missing.
+  #
+  # ```
````
This could probably get a note that the *exact* same value will be used as default, which might have unwanted effects when it is a reference (because then all will point to the same instance).

KCErb replied:
Ooh, I like that, good point. Will do on next commit once I hear a little feedback on the other comments here.

KCErb commented on src/hash.cr:
```diff
@@ -13,6 +13,7 @@ class Hash(K, V)
   @last : Entry(K, V)?
   @block : (self, K -> V)?
 
+  # :nodoc:
```
> I decided to nix the initialize docs since the other two self.new methods are wrappers to it. Let me know if that's misguided.

straight-shoota replied:
No, this overload should also be publicly documented.

Sija replied:
@KCErb ping

KCErb commented on src/hash.cr:
```diff
@@ -23,10 +24,36 @@ class Hash(K, V)
+  # Creates a new empty `Hash` of *initial_capacity* with a *block* that handles missing keys.
+  #
+  # The *initial_capacity* is useful to avoid unnecessary reallocations
+  # of the internal buffer in case of growth. If you have an estimate
```
I just copied and pasted this from the `Array` docs. Not totally sure it's entirely accurate here since I know that bucket size also is used for determining when to rehash. Should I mostly keep it as is but change the voice to avoid "I, me, you" stuff? Should I update the `Array` docs while I'm at it?

KCErb commented on src/hash.cr:
```diff
@@ -23,10 +24,36 @@ class Hash(K, V)
+  # be initialized with that capacity for improved performance.
+  #
+  # NOTE: *initial_capacity* defaults to 11 and an input of < 11 is ignored.
+  # NOTE: `#size` does not reflect capacity as in `Array`.
```
Maybe? It surprised me for just a second before I realized that this makes sense / is obvious, so I decided to make a note in the docs. Happy to remove it.

KCErb commented on src/hash.cr:
```diff
@@ -23,10 +24,36 @@ class Hash(K, V)
     @block = block
   end
 
+  # Creates a new empty `Hash` of *initial_capacity* with a *block* that handles missing keys.
```
:+1:

KCErb commented on src/hash.cr:
```diff
@@ -23,10 +24,36 @@ class Hash(K, V)
+  # of the maximum number of elements a hash will hold, the hash should
+  # be initialized with that capacity for improved performance.
+  #
+  # NOTE: *initial_capacity* defaults to 11 and an input of < 11 is ignored.
```
:+1:

KCErb commented on src/hash.cr:
````diff
@@ -23,10 +24,36 @@ class Hash(K, V)
+  # NOTE: `#size` does not reflect capacity as in `Array`.
+  #
+  # ```
+  # new_hash = Hash(String, Int32).new(50) do |hash, key|
````
I would like to include it since it's one of the args to the method I'm trying to demo. I like the example to show exactly how and where to use the argument. But I agree, it's a pretty pointless example. Is there a way to demo with a meaningful example?

KCErb commented:
Is this one waiting on something from me? I think I've responded to all of the comments. :)

KCErb commented:
@straight-shoota most recent commit addresses the issues you brought up, thanks for the thorough review.

These initializers overlap in interesting ways, so I was a little conflicted on how best to describe the main purpose of each without being repetitive. Ultimately I decided to move the `initialize` definition *beneath* the two `new` overrides which is probably unorthodox. But I think the result pushes this in the right direction.

RX14 commented:
@KCErb please don't reorder here - being repetitive and copy/pasting is fine.

KCErb commented:
Sorry for the slow cycle on this one, I've been much busier than I expected. I restored the order of the initializers and decided to add an example for the first one since it will be the first that a person sees.

KCErb commented:
@straight-shoota thanks for your patience and feedback! I changed the first example to a proc literal as suggested and reworded the note about initial capacity to simply say that if the number of elements is known then this is helpful.

bcardiff commented:
Thanks @KCErb for ~~improving~~ adding docs here. 

Welcome to the contributors list 🎉
